### PR TITLE
Add native causal conv1d full-sequence and decode operations

### PR DESCRIPTION
**Pull request by YangXu1990uiuc**
## Before submitting

- [x] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt).
- [x] I ran `pre-commit run` and committed any formatting changes.
- [x] I added GitHub labels: one `cat-*`, one or more `mod-*`, and one `orig-*` (see [label list](https://github.com/NVIDIA/cudnn-frontend/labels)).
- [ ] I set the **Milestone** and **Projects** fields in the sidebar (required to merge; maintainers can set these for external contributions). The milestone is set to `Frontend 1.29.0`; Project assignment is still required.

## Affected area

FE OSS kernels or CuTeDSL

### Model-level context

On one NVIDIA B200, BF16 batch-one forward+backward proxies measured the following incumbent-to-optimized ratios with the broader optimized feature stack enabled:

| proxy scope | 8K tokens | 16K tokens | 32K tokens |
|---|---:|---:|---:|
| Qwen3.5-9B, 32-layer decoder backbone | **1.1530x** | **1.1722x** | **1.1931x** |
| Qwen3.8-27B, 64-layer decoder backbone | **1.1354x** | **1.1967x** | **1.0898x** |
| GLM-5.3-Flash, 34-layer KDA substack | **1.0732x** | **1.0950x** | **1.0966x** |
| Kimi-K3, official-count feature substack | **1.3567x** | **1.3238x** | **1.2189x** |

These are combined-stack proxy results, not marginal speedups attributable to this PR and not assembled-model throughput. Qwen also changes GDN, SwiGLU, and SDPA; GLM also changes KDA; Kimi also changes KDA and SiTU. Each proxy exercises the full-sequence causal-convolution path landed here.

This PR adds one semantic causal-convolution family:

- `cudnn.ops.causal_conv1d`: native BF16 width-four SiLU full-sequence forward/backward, dense or `cu_seqlens` packed, optional bias and mathematical `W - 1` initial/final state;
- a bias-free BF16-activation/FP32-weight specialization used by GLM linear-attention blocks;
- `cudnn.ops.causal_conv1d_update`: one-token inference with in-place state mutation, identity or SiLU, optional indexed state slots, and `-1` padding rows;
- zero-copy state handoff from full-sequence execution to update; and
- an opt-in FLA 0.5.2 short-convolution compatibility route.

The two native semantic operation entry points are public through `cudnn.ops`. The opt-in public compatibility surface also extends `cudnn.fla.accelerate_fla(targets="short_conv")` and exposes `short_conv_last_path` route telemetry. Compiled plans, schedules, workspaces, and the historically named `*_sm100` packages remain private implementation details. The kernels are independently derived from the convolution and SiLU equations; no external causal-convolution kernel source is included.

## Why

The existing generic cuDNN path accepts a broad convolution contract, but linear-attention short convolution has a narrower and performance-sensitive shape: width four, causal access, optional state, and model-native sequence-major storage. Generic planning and channels-first adaptation can erase an otherwise fast leaf kernel inside the real layer.

The native implementation consumes the model layout directly and specializes the short-width causal work. Full-sequence and update share one state-layout contract, so landing them together guarantees a direct prefill-to-update handoff. This remains a standalone semantic operation rather than a model-topology matcher in the graph API.

## Related issues

Supersedes #797 and #798.

## API and compatibility impact

`cudnn.ops.causal_conv1d` preserves the existing `[B, D, T]` API and adds keyword-only packed/state arguments. Existing dense/stateless calls retain the generic cuDNN route when they do not match the optimized contract. The native route covers BF16 width-four SiLU with optional BF16 bias, plus dense bias-free FP32 weights. `seq_idx` is reserved and currently declines explicitly.

Dense state is `[B, D, W - 1]`; packed state is `[N, D, W - 1]`, where `N = len(cu_seqlens) - 1`.

`cudnn.ops.causal_conv1d_update(x, conv_state, weight, bias=None, activation=None, *, cache_seqlens=None, conv_state_indices=None)` accepts BF16 one-token input `[N, D]`, mutates BF16 state `[S, D, L]` in place, and returns `[N, D]`. Width four is supported with `L` equal to three or four. The operation is inference-only; `cache_seqlens` is reserved and must be `None` or omitted.

Runtime correctness has been exercised on SM80, SM89, SM90, SM100, SM103, and SM120; SM86, SM87, SM110, and SM121 are compile-validated. Architecture is a schedule-selection dimension, not a semantic API gate. Performance claims in this PR are B200-only. The implementation requires `nvidia-cutlass-dsl >= 4.7.0`; users select the package variant matching their CUDA Toolkit.

## Testing

Current review head: `96c524644`. The latest commit changes architecture wording only; kernel code remains unchanged from `87f68dad3`.

- `pre-commit run --from-ref upstream/develop --to-ref HEAD`: passed.
- exact current-tree guardword scan over all 285 PR-changed paths: **285/285 passed**; Python compilation and diff checks also passed.
- focused host-visible contract suite with `-m L0`: `173 passed, 1 deselected`; the isolated contract module with `-m L1`: `1 passed, 42 deselected`.
- private-package lazy-import smoke: passed; requesting forward does not import autograd/backward.
- Python 3.9 grammar parse over repository Python files: passed.
- the three post-kernel review commits (`26cad8e03`, `658b27632`, `96c524644`) change documentation, import/API metadata contracts, test scheduling, and neutral architecture wording; kernel code is unchanged from `87f68dad3`.
- kernel-identical NVIDIA B200 L1 suite at `26cad8e03` across full-sequence forward/backward, update, and the prefill-to-update seam: `33 passed, 187 deselected`.

Focused host command:

```bash
python -m pytest -q -m L0 \
  test/python/fe_api/causal_conv1d_bulk/test_causal_conv1d_bulk_backward_contract_unit.py \
  test/python/fe_api/causal_conv1d_bulk/test_causal_conv1d_bulk_contract_unit.py \
  test/python/fe_api/causal_conv1d_bulk/test_causal_conv1d_semantic_api_unit.py \
  test/python/fe_api/causal_conv1d_update/test_causal_conv1d_update_benchmark_contract_unit.py \
  test/python/fe_api/causal_conv1d_update/test_causal_conv1d_update_contract_unit.py \
  test/python/linear_attention/test_fla_short_conv_shim_unit.py
```

B200 correctness command:

```bash
python -m pytest -q -m L1 \
  test/python/fe_api/causal_conv1d_bulk \
  test/python/fe_api/causal_conv1d_update \
  test/python/linear_attention/test_fla_short_conv_shim_unit.py
```

### B200 operator evidence

At the unchanged kernel head `87f68dad3553cb8f325af8038c77567267bff5fb`:

| scope | representative workload and timing boundary | result |
|---|---|---:|
| full-sequence forward | dense `T=8192, D=2048`; CUDA-event direct calls | prepared FE **26.618 us** vs direct FLA **64.922 us**, **2.439x**; allocation-inclusive FE **44.058 us** vs public FLA **89.146 us**, **2.023x** |
| backward | dense `B=1, T=8192, D=2048`; warmed CUDA Graph replay | native **28.032 us** vs FLA **130.592 us**, **4.659x** |
| update | `N=1, D=8192`; warmed CUDA Graph replay | FE **7.840 us** vs FLA **7.616 us**; order-sensitive and effectively tied |

Reproduction commands:

```bash
python test/python/fe_api/causal_conv1d_bulk/benchmark_causal_conv1d_bulk_sm100.py --tokens 8192 --channels 2048
python benchmark/causal_conv1d_bulk_bwd_perf.py --tokens 8192 --channels 2048
python benchmark/causal_conv1d_update_sm100.py --shape 1x8192
```

Compilation, capture, and allocation are outside the backward/update replay intervals. The full-sequence comparison uses equal contiguous `[1, T, D]` BF16 layouts. Update is reported as parity; its primary value here is native state compatibility rather than a speedup claim.

The focused NVIDIA B200 correctness suite at `87f68dad3` reported `88 passed, 1 warning`, covering public full-sequence forward/backward, update, the state seam, real FLA compatibility, non-default streams, CUDA Graph replay, and `torch.compile`/FakeTensor contracts. The internal `python_tests,frost` pipeline was also run for that SHA.

## Summary by CodeRabbit

* **New Features**
  * Added full-sequence causal convolution for dense and packed inputs, with bias, state handling, and gradient support.
  * Added single-token causal-convolution updates with mutable state, activations, bias, and indexed routing.
  * Added optimized BF16 execution paths for supported NVIDIA GPUs.
  * Added optional FLA short-convolution acceleration with safe fallback behavior.

* **Documentation**
  * Documented causal-convolution, decode-update, and FLA short-convolution APIs, limitations, and usage.

* **Tests & Benchmarks**
  * Added extensive correctness, compatibility, smoke-test, and performance coverage.

### CODE REVIEW

**Comment by YangXu1990uiuc** [comment]
Model-callsite closure update at head 5ef325702:

- The public causal_conv1d route now accepts the exact GLM-5.3 contract: BF16 activations, bias-free FP32 width-four weights, BF16 outputs, and FP32 weight gradients. Other mixed-dtype/bias combinations remain fail-closed.
- The implementation commit 077fa3d72 has the same stable patch-id as the clean composite source used by the GLM model proxy; docs are updated in 5ef325702.
- Focused host contract tests: 112 passed. ComputeLab B200 job 4044080, serialized step 4044080.141: three exact mixed-weight forward/backward/semantic tests passed against the independent FP32 oracle. Pre-commit passed on every changed file.

In the exact Transformers 5.16.1 GLM-5.3 KDA callsite, this route composes with native KDA in the 34-layer KDA substack. The controlled all-on fprop+bprop proxy measured 1.0732x / 1.0950x / 1.0966x at 8K / 16K / 32K versus FLA 0.5.2 KDA plus causal-conv1d 1.7.0. Those are combined KDA+conv substack results, not a marginal speedup attributed solely to this PR.

File: docs/fe-oss-apis/overview.md [hunk 1]
```diff
@@ -9,6 +9,7 @@ The GEMM CuTeDSL APIs are type-erased and torch-lazy: torch is imported only whe
 - **proj_rope_mxfp8**: JAX eager support on both input paths with `w_out_in=True` (the transposed [in, out] weight view is torch-only), plus the `jax.jit`-compatible `gemm_proj_rope_mxfp8_jax_sm100` entry point.
 
 This folder documents the Python FE APIs implemented under `python/cudnn`. For details on currently implemented operations, see:
+- [Bulk Causal Conv1d Forward (SM100)](causal_conv1d_bulk_contract.md)
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply docs/fe-oss-apis/overview.md]: Intentional after the semantic API rewrite: the docs index the model-facing cudnn.ops.causal_conv1d API. Bulk forward, backward, and autograd classes are private implementation seams and deliberately absent from __all__.

File: docs/fe-oss-apis/causal_conv1d_bulk_contract.md [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+Every tensor is CUDA-resident. Data tensors are BF16 and must have a 16-byte
+aligned base pointer; `cu_seqlens` is int32 with a 4-byte aligned base pointer.
+This API requires `nvidia-cutlass-dsl>=4.7.0`; the frontend package's broader
+`cutedsl` extra deliberately retains its 4.5.0 floor for unrelated APIs.
+The first slice is inference-only and rejects requires-grad inputs while grad
+mode is enabled. There is no implicit dtype/layout conversion or fallback.
+On SM100, SM103, SM110, SM120, and SM121, channel extents divisible by eight
+use an aligned 128-bit channel-vector fast path with packed FP32 arithmetic.
+SM80 through SM90, and every positive channel extent not divisible by eight,
+use a predicated scalar path with the same public semantics. Only SM100/B200
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply docs/fe-oss-apis/causal_conv1d_bulk_contract.md]: Superseded by the semantic API rewrite. The removed bulk-contract wording no longer exists; supported capabilities remain enumerated explicitly in _causal_conv1d_bulk_arch.py.

File: benchmark/causal_conv1d_bulk_bwd_smoke.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    fla_dw_diff = fla_dw.float() - reference_dw
+    fla_metrics = {
+        "dx_max_abs": float(fla_dx_diff.abs().max()),
+        "dx_rel_l2": float(torch.linalg.vector_norm(fla_dx_diff) / torch.linalg.vector_norm(reference_dx)),
+        "dw_max_abs": float(fla_dw_diff.abs().max()),
+        "dw_rel_l2": float(torch.linalg.vector_norm(fla_dw_diff) / torch.linalg.vector_norm(reference_dw)),
+    }
+
+    rows = {}
+    for schedule in ("t32", "t64", "t128", "t64-partial"):
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply benchmark/causal_conv1d_bulk_bwd_smoke.py]: Retained. t32 is a declared entry in _SCHEDULES, and the constructor accepts auto plus every _SCHEDULES key. The standalone performance harness also uses t32. Removing it would drop a valid fixed schedule.

File: python/cudnn/causal_conv1d_bulk_sm100/backward_kernel.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+        )
+
+        descriptor = cutlass.Int32(0)
+        sequence = cutlass.Int32(0)
+        while sequence < num_sequences:
+            sequence_start = cutlass.Int64(cu_seqlens[sequence])
+            sequence_end = cutlass.Int64(cu_seqlens[sequence + cutlass.Int32(1)])
+            inline_ptx("trap;", predicate=sequence_start < cutlass.Int64(0))
+            inline_ptx("trap;", predicate=sequence_end <= sequence_start)
+            inline_ptx("trap;", predicate=sequence_end > total_tokens)
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply python/cudnn/causal_conv1d_bulk_sm100/backward_kernel.py]: Fixed in 89d035d3b. Both backward_kernel.py and causal_conv1d.md now document the device trap, sticky asynchronous CUDA failure, and need to discard the affected context.

File: docs/fe-oss-apis/causal_conv1d_bulk_contract.md [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+y = result["output_tensor"]
+final_state = result["final_state_tensor"]
+```
+
+When `output_final_state=False`, `final_state_tensor` is a CUDA BF16 sentinel
+with shape `(0,)`. An FLA adapter must translate that sentinel to `None`.
+
+The lower-level class API follows the common FE-OSS lifecycle: construct it
+from representative PyTorch tensors (metadata-only `TensorDesc` inputs are not
+accepted), call `check_support()`, call `compile()` once,
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply docs/fe-oss-apis/causal_conv1d_bulk_contract.md]: Superseded by the semantic API rewrite. The model-facing public API is cudnn.ops.causal_conv1d; the Tensor-sample bulk classes are private implementation and test seams with an empty __all__.

File: test/python/fe_api/causal_conv1d_bulk/test_causal_conv1d_bulk_backward_contract_unit.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+    monkeypatch.setattr(backward, "cutedsl_state", lambda: (True, None))
+
+    with pytest.raises(ValueError, match="Bias"):
+        api.check_support()
+
+
+def test_backward_rejects_non_tensor_bias():
+    _, _, api_class, _ = _load_backward()
+    x = torch.zeros(1, 4, 8, dtype=torch.bfloat16)
+    with pytest.raises(TypeError, match="sample_bias must be a torch.Tensor or None"):
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply test/python/fe_api/causal_conv1d_bulk/test_causal_conv1d_bulk_backward_contract_unit.py]: Fixed in 89d035d3b. Both patterns now use raw strings with an escaped literal period; additional same-file patterns were fixed and targeted Ruff RUF043 passes.

File: test/python/fe_api/causal_conv1d_bulk/test_causal_conv1d_bulk_sm100.py [hunk 7]
```diff
@@ -0,0 +1,603 @@
...
+    torch.testing.assert_close(result["output_tensor"], expected, atol=3e-2, rtol=3e-2)
+    _assert_state_bits_equal(result["final_state_tensor"], expected_final)
+    _assert_state_bits_equal(initial_state, initial_state_before)
+
+    # N is part of the cache key but T is symbolic. Rebinding this cached packed
+    # signature to fewer tokens than positive sequences must fail on the host,
+    # before the device-side metadata validator can trap.
+    too_short_x = torch.randn(1, 3, channels, device="cuda", dtype=torch.bfloat16)
+    too_short_cu = torch.tensor((0, 1, 2, 3, 3, 3), device="cuda", dtype=torch.int32)
+    with pytest.raises(ValueError, match="cannot exceed total_T"):
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply test/python/fe_api/causal_conv1d_bulk/test_causal_conv1d_bulk_sm100.py]: Retained the original expectation based on an executed focused test. The public wrapper raises Packed N=5 cannot exceed total_T=3 before the lower-level execute guard. Changing the match to runtime total_T made this exact test fail; restoring cannot exceed total_T passes.

File: python/cudnn/causal_conv1d_bulk_sm100/backward.py [hunk 8]
```diff
@@ -0,0 +1,572 @@
...
+    "t64": _BulkBwdSchedule(256, 64, _DWEIGHT_ATOMIC),
+    "t128": _BulkBwdSchedule(256, 128, _DWEIGHT_ATOMIC),
+    "t64-partial": _BulkBwdSchedule(256, 64, _DWEIGHT_PARTIAL),
+}
+
+
+def select_bulk_bwd_schedule(total_tokens: int) -> str:
+    """Choose the small schedule table measured for the dense prototype."""
+
+    return "t128" if total_tokens >= 16384 else "t64"
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply python/cudnn/causal_conv1d_bulk_sm100/backward.py]: Fixed in 89d035d3b. Module docs and select_bulk_bwd_schedule now state that t64 and t128 use FP32 atomics and are not bitwise reproducible even under PyTorch deterministic mode; t64-partial is documented as the reproducible route.

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run frost,python_tests

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run python_tests,frost

**Comment by YangXu1990uiuc** [comment]
@coderabbitai review

**Comment by YangXu1990uiuc** [comment]
@coderabbitai review

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run python_tests,frost

File: docs/fe-oss-apis/fla.md [hunk 9]
```diff
@@ -61,13 +107,15 @@ per-thread state.
...
 pip install flash-linear-attention==0.5.2
 pip install "nvidia-cudnn-frontend[cutedsl]"
 ```
 
-The `cutedsl` extra supplies the optional dependencies required by the fused
-GEMM path used by the native `gated_mlp` target.
+The `cutedsl` extra supplies the optional CUTLASS DSL and CUDA Python
+dependencies required by the native `gated_mlp` and `short_conv` targets. To
+require the native `short_conv` route instead of its typed fallback, also
+ensure `nvidia-cutlass-dsl[cu13]>=4.7` is installed.
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply docs/fe-oss-apis/fla.md]: Addressed in 26cad8e03: the docs require `nvidia-cutlass-dsl>=4.7` and tell users to select the package variant matching their CUDA Toolkit.

File: docs/fe-oss-apis/causal_conv1d_update.md [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+  target
+- performance-characterized GPU: B200 SM100
+- `x`: BF16 `[N, D]` with strides `(ld, 1)`; compact `ld == D` accepts every
+  channel count, and padded `ld > D` requires `ld % 8 == 0`
+- `weight`: contiguous BF16 `[D, 4]`
+- `conv_state`: BF16 `[S, D, L]` with `L` equal to 3 or 4, updated in place;
+  compact storage is accepted for both lengths, and `L=3` additionally accepts
+  the channel-last `(3 * D, 1, D)` stride returned by full-sequence prefill
+- `bias`: optional contiguous BF16 `[D]`
+- `cache_seqlens`: must be omitted
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply docs/fe-oss-apis/causal_conv1d_update.md]: Addressed in 26cad8e03: the docs now state that `cache_seqlens` may be `None` or omitted.

File: docs/fe-oss-apis/causal_conv1d.md [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+tensor and state semantics match the commonly consumed `causal-conv1d`
+interface:
+
+- `x[B, D, T]`
+- `weight[D, W]` and optional `bias[D]`
+- reserved `seq_idx[B, T]` compatibility keyword (currently only `None` is
+  accepted)
+- optional packed `cu_seqlens[N + 1]` (mutually exclusive with `seq_idx`)
+- dense and `seq_idx` state: optional `initial_states[B, D, W - 1]` and
+  returned `final_states[B, D, W - 1]`
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply docs/fe-oss-apis/causal_conv1d.md]: Addressed in 26cad8e03: the state contract now documents dense and packed modes only; `seq_idx` remains explicitly reserved.

File: test/python/fe_api/causal_conv1d_bulk/test_causal_conv1d_bulk_backward_contract_unit.py [hunk 12]
```diff
@@ -0,0 +1,586 @@
...
+    assert vec4.tokens_per_cta == 360
+    assert vec4.tiles_per_sequence == 46
+
+
+def test_v2_cpasync_planner_is_safe_for_arbitrary_t_d_and_sm(monkeypatch):
+    backward, *_ = _load_backward()
+    for sequence_length in range(64, 32769):
+        for n_channels in (512, 1024, 2048, 4096, 8192, 16384):
+            channel_ctas = n_channels // backward._VEC2_CPASYNC_CHANNELS_PER_CTA
+            for sm_count in (40, 60, 80, 100, 120, 148, 160):
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply test/python/fe_api/causal_conv1d_bulk/test_causal_conv1d_bulk_backward_contract_unit.py]: Addressed in 26cad8e03: L0 now samples the planner space while retaining boundary cases instead of running the exhaustive 1.37M-case sweep.

File: python/cudnn/causal_conv1d_bulk_sm100/__init__.py [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+deliberately absent from the package's public export list; model code should
+use ``cudnn.ops``.
+"""
+
+from .api import CausalConv1dBulkFwdSm100, causal_conv1d_bulk_fwd_wrapper_sm100
+from .autograd import CausalConv1dBulkAutogradPrototype
+from .backward import (
+    CausalConv1dBulkBwdPrototype,
+    compile_causal_conv1d_bulk_bwd_prototype,
+)
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply python/cudnn/causal_conv1d_bulk_sm100/__init__.py]: Addressed in 26cad8e03: forward import no longer loads autograd/backward. The backend package remains private with an empty `__all__`; the supported public entry is `cudnn.ops.causal_conv1d`.

File: test/python/fe_api/causal_conv1d_bulk/test_causal_conv1d_bulk_backward_sm100.py [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+def _reference(
+    x: torch.Tensor,
+    weight: torch.Tensor,
+    bias: torch.Tensor | None,
+    offsets: tuple[int, ...] | None,
+) -> torch.Tensor:
+    if offsets is None:
+        return _dense_reference(x, weight, bias)
+    return torch.cat(
+        [_dense_reference(x[:, start:end], weight, bias) for start, end in zip(offsets[:-1], offsets[1:], strict=True)],
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply test/python/fe_api/causal_conv1d_bulk/test_causal_conv1d_bulk_backward_sm100.py]: Addressed in 26cad8e03: removed the Python-3.10-only `zip(strict=...)` use and verified Python 3.9 grammar compatibility.

File: test/python/fe_api/causal_conv1d_update/test_causal_conv1d_update_sm100.py [hunk 15]
```diff
@@ -0,0 +1,529 @@
...
+    os._exit(0)
+
+print(f"MISSING_DEVICE_FAILURE:{case}", flush=True)
+os._exit(9)
+"""
+
+
+@pytest.mark.L1
+@pytest.mark.parametrize("case", ["below_padding", "out_of_range", "duplicate"])
+def test_invalid_state_indices_fail_closed_in_fresh_process(case):
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply test/python/fe_api/causal_conv1d_update/test_causal_conv1d_update_sm100.py]: Addressed in 26cad8e03: level markers are explicit per test, so the fresh-process cases are L1-only and no longer inherit L0.

File: test/python/fe_api/causal_conv1d_bulk/test_causal_conv1d_bulk_contract_unit.py [hunk 16]
```diff
@@ -0,0 +1,370 @@
...
+        shape=(1, 2, 8),
+        stride=(16, 8, 1),
+        stride_order=(2, 1, 0),
+        device=torch.device("cuda"),
+    )
+    weight = torch.zeros(8, 4, dtype=torch.bfloat16)
+    output = torch.zeros(1, 2, 8, dtype=torch.bfloat16)
+
+    with pytest.raises(TypeError, match=r"sample_x must be a torch.Tensor, got TensorDesc"):
+        api_class(sample_x, weight, output)
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply test/python/fe_api/causal_conv1d_bulk/test_causal_conv1d_bulk_contract_unit.py]: Addressed in 26cad8e03: metadata-only `TensorDesc` samples are accepted and covered by an observable support test.

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run frost,python_tests

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run frost,python_tests

**Comment by YangXu1990uiuc** [comment]
@coderabbitai review

**Comment by swalters22** [comment]
I spent some time w/codex on this; as near as we can tell, there are no SS or P0 issues.  There are a few P1 issues to consider, but not blocking if followed up upon later.

**[P1] Reject `final_states_out` aliases before they corrupt saved inputs**
The public validator accepts `final_states_out` after checking only its shape, dtype, and device. It does not ensure that the output buffer is disjoint from inputs:

**[P1] Honor `torch.use_deterministic_algorithms` instead of silently using atomic backward**
The public operation silently selects an atomic dWeight schedule while strict deterministic mode is enabled.

In each case an adversarial exercise of the issue can be made.

Codex is adamant that this could be refactored into a smaller PR, but I experimented less with that FWIW.

**Review by swalters22 — APPROVED** [review]
See discussion comment re: P1 issues + refactoring.  Approving as no blocking issues were observed during testing.

**Comment by YangXu1990uiuc** [comment]
@swalters22 thanks — both P1s are addressed at head `470c699b7` (also rebased onto develop; the only conflict was `docs/fe-oss-apis/overview.md`).

**[P1] `final_states_out` aliasing.** The public validator now rejects a `final_states_out` that shares memory with `x`, `weight`, `bias`, `cu_seqlens`, or `initial_states` (`ValueError`). The check uses addressed byte spans of the (possibly strided) views rather than `Storage` identity, so DLPack-wrapped buffers are covered too. Host tests: `x[..., :3]`, an `as_strided` view of `weight`, and `initial_states` itself are all rejected; disjoint slices of one storage still pass.

**[P1] `torch.use_deterministic_algorithms`.** The torch flag now keys the exact-shape training plan cache and reaches the backward backend. With deterministic mode on, bias-free calls select the partial-dweight schedule (unique FP32 partials + fixed-order reduction; the streaming schedules were already partial). Calls with `bias` have no deterministic dbias route: they raise `RuntimeError`, or warn and run atomically under `warn_only=True` — same semantics as torch and as the CSA compressor precedent in this repo. An explicitly requested atomic schedule under deterministic mode is rejected the same way. Default (non-deterministic) behavior and measured schedules are unchanged.

Docs updated accordingly; the FP32-weight dispatch comment/docs now state the tested capability instead of naming a model. Host-side unit suites: 185 passed.

On "could be a smaller PR": agreed in principle, but the history is already a single squashed commit, so splitting is no longer practical; follow-ups will land as separate PRs.

<sub>note to self: claude::c21097e1-70b0-4dab-b38c-343d85ba2e22 — "codex/kernel-factory PR batch design audit (797–814 + 835)" · cwd /home/scratch.yanxu_libs/cudnn_frontend</sub>

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run python_tests,frost
